cdn/signedurls/signurl.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2018 Google Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+// Command signedurls contains helper function to sign URLs.
+package main
+
+import (
+	"crypto/hmac"
+	"crypto/sha1"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+)
+
+// [START example]
+
+// SignURL creates a signed URL for an endpoint on Cloud CDN. sourceURL must
+// start with "https://" and should not have the Expires, KeyName, or Signature
+// query parameters. key should be in raw form (not base64url-encoded).
+func SignURL(sourceURL, keyName string, key []byte, expiration time.Time) string {
+	u := sourceURL
+	sep := "?"
+	if strings.Contains(u, "?") {
+		sep = "&"
+	}
+	u += sep
+	u += fmt.Sprintf("Expires=%d", expiration.Unix())
+	u += fmt.Sprintf("&KeyName=%s", keyName)
+
+	sig := base64urlencode(hmacSHA1(u, key))
+	u += fmt.Sprintf("&Signature=%s", sig)
+	return u
+}
+
+func hmacSHA1(msg string, key []byte) []byte {
+	mac := hmac.New(sha1.New, key)
+	mac.Write([]byte(msg))
+	return mac.Sum(nil)
+}
+
+func base64urlencode(v []byte) string {
+	s := base64.StdEncoding.EncodeToString(v)
+	s = strings.Replace(s, "+", "-", -1)
+	s = strings.Replace(s, "/", "_", -1)
+	return s
+}
+
+func base64urldecode(s string) ([]byte, error) {
+	s = strings.Replace(s, "-", "+", -1)
+	s = strings.Replace(s, "_", "/", -1)
+	b, err := base64.StdEncoding.DecodeString(s)
+	return b, err
+}
+
+// readKeyFile reads the base64url-encoded key file and decodes it.
+func readKeyFile(path string) ([]byte, error) {
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read key file: %+v", err)
+	}
+	decoded, err := base64urldecode(string(b))
+	if err != nil {
+		return nil, fmt.Errorf("failed to base64url decode: %+v", err)
+	}
+	return decoded, nil
+}
+
+func main() {
+	key, err := readKeyFile("/path/to/key")
```

Comment:
This program isn't really runnable as-is. Could you make the key file a parameter?